File: Data Structures & Algorithms/target-sum/submission-13.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        
        def dfs(i, cur): 
            if i == len(nums): 
                return 1 if cur == target else 0 

            take = dfs(i + 1, cur + nums[i]) 
            minus = dfs(i + 1, cur - nums[i]) 

            return take + minus  

        return dfs(0, 0)

File: Data Structures & Algorithms/combination-target-sum/submission-13.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {  
        let res = []
        function dfs(j, cur, curSum) {
            if (curSum === target) {
                res.push([...cur]) 
                return
            }  

            if (curSum > target) {
                return
            }

            for (let i = j; i < nums.length; i ++) {       
                cur.push(nums[i]) 
                dfs(i, cur, curSum + nums[i]) 
                cur.pop()
            }
        } 

        const cur = [] 
        dfs(0, cur, 0) 
        return res
    }
}
